lib/grpcxx/uv/server.cpp

```diff
@@ -54,21 +62,29 @@ void server::conn_cb(uv_stream_t *stream, int status) {
 	s->conn(stream);
 }
 
-void server::prepare(std::string_view ip, int port) {
-	sockaddr_storage addr;
-	if (uv_ip4_addr(ip.data(), port, reinterpret_cast<sockaddr_in *>(&addr)) != 0 &&
-		uv_ip6_addr(ip.data(), port, reinterpret_cast<sockaddr_in6 *>(&addr)) != 0) {
-		throw std::runtime_error(std::string(ip) + " is not a valid IPv4 or IPv6 address");
+void server::listen() {
+	if (auto r = uv_listen(reinterpret_cast<uv_stream_t *>(&_handle), TCP_LISTEN_BACKLOG, conn_cb);
+		r != 0) {
+		throw std::runtime_error(
+			std::string("Failed to listen for connections: ") + uv_strerror(r));
 	}
+}
 
-	if (auto r = uv_tcp_bind(&_handle, reinterpret_cast<const sockaddr *>(&addr), 0); r != 0) {
-		throw std::runtime_error(std::string("Failed to bind to tcp address: ") + uv_strerror(r));
-	}
+uv_loop_t *server::listen(uv_os_sock_t &&sock) {
+	open(sock);
+	listen();
 
-	start_listening();
+	return _loop;
 }
 
-void server::prepare(uv_os_sock_t sock) {
+uv_loop_t *server::listen(std::string_view ip, int port) {
+	bind(std::move(ip), port);
+	listen();
+
+	return _loop;
+}
+
+void server::open(uv_os_sock_t sock) {
```

Comment:
Isn't this conceptually moved too? Any further operation should happen through the `_handle`, not the naked `sock` anymore, right?

Reply:
🤔 true, I thought since this is a private method we don't really need to move but agree it's better to make it explicit. I'll change.